File: solutions/Two Pointers/125. Valid Palindrome.java
//https://leetcode.com/problems/valid-palindrome

// First Solution - Time complexity: O(n) - Space complexity: O(n)
class Solution {
    public boolean isPalindrome(String s) {
        // Remove spaces and all the stuff that is not a letter/number
        s = s.toLowerCase().replaceAll("[^a-z0-9]", "");

        Deque<Character> queue = new ArrayDeque<>();
        Deque<Character> stack = new ArrayDeque<>();
        
        for (char c : s.toLowerCase().toCharArray()) {
            queue.offer(c);
            stack.push(c);
        }

        while (!queue.isEmpty())
            if (queue.poll() != stack.pop()) 
                return false;

        return true;
    }
}

// Second Solution - Time complexity: O(n) - Space complexity: O(1)
class Solution {
    public boolean isPalindrome(String s) {
        // Remove spaces and all the stuff that is not a letter/number
        s = s.toLowerCase().replaceAll("[^a-z0-9]", "");

        int left = 0, right = s.length() - 1;

        while(left < right) { 
            if(s.charAt(left) != s.charAt(right))
                return false;

            left++;
            right--;
        }
        
        return true;
    }
}

File: solutions/Array-String/125. Valid Palindrome.py
#https://leetcode.com/problems/valid-palindrome

import re

# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """

        s = self._sanitize(s)

        left, right = 0, len(s) - 1

        while left < right:
            if s[left] != s[right]:
                return False
            
            right -= 1
            left += 1

        return True

    # Space Complexity: O(n)
    def _sanitize(self, s):
        s = re.sub(r'[^a-zA-Z0-9]', '', s)

        """
        s = s.replace(" ", "")
        s = s.replace(",", "")
        s = s.replace(".", "")
        s = s.replace(":", "")
        s = s.replace("/", "")
        s = s.replace("@", "")
        """

        return s.lower()
        

#############################

# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """

        left, right = 0, len(s) - 1

        while left < right:
            while not s[left].isalnum() and left < right:
                left += 1
            while not s[right].isalnum() and left < right:
                right -= 1

            if s[left].lower() != s[right].lower():
                return False
            
            right -= 1
            left += 1

        return True
        